sphinx/builders/linkcheck.py

```diff
@@ -283,13 +285,23 @@ def __init__(self, config: Config,
         self.allowed_redirects = config.linkcheck_allowed_redirects
         self.retries: int = config.linkcheck_retries
         self.rate_limit_timeout = config.linkcheck_rate_limit_timeout
+        self.allow_unauthorized = config.linkcheck_allow_unauthorized
 
         self.user_agent = config.user_agent
         self.tls_verify = config.tls_verify
         self.tls_cacerts = config.tls_cacerts
 
         self._session = requests._Session()
 
+        if config.linkcheck_allow_unauthorized is not None:
```

Comment:
I think it's better to just keep `linkcheck_allow_unauthorized = True` by default. At least you don't have the `None` case, but you could still emit a warning if `linkcheck_allow_unauthorized` is `False`.

```suggestion
        if not config.linkcheck_allow_unauthorized:
```

I would suggest emitting the warning before opening the session also (namely the sooner the better).

Reply:
Ok, that seems fair.  That was my original plan (to leave the default as `True`) - I changed my mind [here](https://github.com/sphinx-doc/sphinx/pull/11684#discussion_r1334501185) because I figured that we should warn about settings that will be removed regardless of what they are configured to...

...but I suppose it's also true that a deployment that has opted-in (configured `True` explicitly) doesn't really need to know that the setting will be removed, so long as it becomes the default at removal-time.

Basically: I'm looking for reassurance / a double-check that a default of `True` without a warning is OK.  I think it is, and will begin to make that adjustment.

Reply:
Sorry, I made a mistake in that middle sentence:

> ...but I suppose it's also true that a deployment that has opted-in (configured `True` explicitly) doesn't really need to know that the setting will be removed, so long as it becomes the default at removal-time.

That should read: "opted in (configured `False` explicitly)".

Booleans and opt-in/out breaking my brain.

Reply:
No, I'm still having trouble figuring out what to do here.  I believe the following are facts:

  * We don't want to add warnings where deployments wouldn't currently see them, because warnings are often treated as errors and we don't want unnecessary breakage.
  * We want the default behaviour to be the existing "HTTP 401 is acceptable" logic.
  * We want to warn about the configuration setting.
  * The configuration setting only has an effect if it is configured to `False` (change to the planned future behaviour).

If a deployment has opted-in to the future behaviour, do they really need a deprecation warning about the setting?

And if a deployment as not opted-in, how do we warn people without introducing possible warning-as-error breakage?

Reply:
> We want to warn about the configuration setting.

Maybe that's an incorrect assumption here?

(I was the one who [decided that a deprecation warning could make sense](https://github.com/sphinx-doc/sphinx/pull/11684#discussion_r1333134731).  maybe I just created a lot of confusion and extra work for myself)

Reply:
> We don't want to add warnings where deployments wouldn't currently see them, because warnings are often treated as errors and we don't want unnecessary breakage.

Indeed.

> We want the default behaviour to be the existing "HTTP 401 is acceptable" logic.

Indeed.

> We want to warn about the configuration setting.

Only if the user changes it, namely only if the default value is changed. In particular, the default value should reflect "HTTP 401 is acceptable". As such, `linkcheck_allow_unauthorized` must be set to `True` by default and can be set to `False` to reflect "401 is not ok".

> The configuration setting only has an effect if it is configured to False (change to the planned future behaviour).

Indeed.

> If a deployment has opted-in to the future behaviour, do they really need a deprecation warning about the setting?

I think so. Btw, I think that deprecation warnings (emitted using the `warnings` module and not using a logger) are *not* treated as errors even with `-W`.

> And if a deployment [h]as not opted-in, how do we warn people without introducing possible warning-as-error breakage?

We don't warn them, namely they will keep their current behaviour. If they don't use the feature, they don't need to know that it will be removed, right? (~~the less they know, the better~~) If warnings are treated as errors, a deprecation warning will still not be considered as an error, so we can emit one when `linkcheck_allow_unauthorized` is set manually. Then, if a 401 occurs, the logger's warning will be treated as an error and we're good.

Reply:
Perfect, thank you @picnixz - that should allow me to apply the remaining code review suggestions.  I'll push a commit or two with those changes soon.

Reply:
> I think so. Btw, I think that deprecation warnings (emitted using the warnings module and not using a logger) are not treated as errors even with -W.

I've been double-checking this, and it looks like `DeprecationWarning` (and subclasses, like `RemovedInSphinx80Warning`) _are_ still converted to exceptions by `-Werror` (and equivalents)).

Reply:
Wait, how come? I've tried locally and using sphinx option `-W` (not Python `-W`!).

Reply:
Agh!  Ok.  I had been reading the Python `warnings` documentation and incorrectly assumed that was the `-W` you were referencing; I didn't check the `sphinx-build` options until now.  Ok.

After checking the `sphinx-build` `-W` option: you're absolutely correct, yep - warnings can be emitted with that flag enabled, and (as expected) the program continues.